modules/stochastic_tools/src/surrogates/LibtorchSimpleNNTrainer.C

```diff
@@ -0,0 +1,199 @@
+  params.addRangeCheckedParam<unsigned int>(
+      "num_epochs", 1, "0<num_epochs", "Number of training epochs.");
+  params.addRangeCheckedParam<Real>(
+      "rel_loss_tol",
+      0,
+      "0<=rel_loss_tol<=1",
+      "The relative loss where we stop the training of the neural net.");
+  params.addParam<std::vector<unsigned int>>(
+      "num_neurons_per_layer", std::vector<unsigned int>(), "Number of neurons per layer.");
+  params.addParam<std::string>(
+      "filename", "net.pt", "Filename used to output the neural net parameters.");
+  params.addParam<bool>("read_from_file",
+                        false,
+                        "Switch to allow reading old trained neural nets for further training.");
+  params.addParam<Real>("learning_rate", 0.001, "Learning rate (relaxation).");
+  params.addRangeCheckedParam<unsigned int>(
+      "print_epoch_loss",
+      0,
+      "0<=print_epoch_loss",
+      "Epoch training loss printing. 0 - no printing, 1 - every epoch, 10 - every 10th epoch.");
+  params.addParam<unsigned int>(
+      "seed", 11, "Random number generator seed for stochastic optimizers.");
+
+  return params;
+}
+
+LibtorchSimpleNNTrainer::LibtorchSimpleNNTrainer(const InputParameters & parameters)
+  : SurrogateTrainer(parameters),
+    _sampler_row(getSamplerData()),
+    _response(getTrainingData<Real>(getParam<ReporterName>("response"))),
+    _num_batches(getParam<unsigned int>("num_batches")),
+    _num_epocs(getParam<unsigned int>("num_epochs")),
+    _rel_loss_tol(getParam<Real>("rel_loss_tol")),
+    _num_neurons_per_layer(declareModelData<std::vector<unsigned int>>(
+        "num_neurons_per_layer", getParam<std::vector<unsigned int>>("num_neurons_per_layer"))),
+    _num_hidden_layers(
+        declareModelData<unsigned int>("num_hidden_layers", _num_neurons_per_layer.size())),
+    _filename(getParam<std::string>("filename")),
+    _read_from_file(getParam<bool>("read_from_file")),
+    _learning_rate(getParam<Real>("learning_rate")),
+    _print_epoch_loss(getParam<unsigned int>("print_epoch_loss"))
+#ifdef TORCH_ENABLED
+    ,
+    _nn(declareModelData<std::shared_ptr<StochasticTools::LibtorchSimpleNeuralNet>>("nn"))
+#endif
+{
+  // We check if MOOSE is compiled with torch, if not this throws an error
+  StochasticToolsApp::requiresTorch(*this);
+
+#ifdef TORCH_ENABLED
+  // Fixing the RNG seed to make sure every experiment is the same.
+  // Otherwise sampling / stochastic gradient descent would be different.
+  torch::manual_seed(getParam<unsigned int>("seed"));
+#endif
+}
+
+void
+LibtorchSimpleNNTrainer::preTrain()
+{
+  // Resize to number of sample points
+  _flattened_data.resize(_sampler.getNumberOfLocalRows() * _sampler.getNumberOfCols());
+  _flattened_response.resize(_sampler.getNumberOfLocalRows());
+}
+
+void
+LibtorchSimpleNNTrainer::train()
+{
+  unsigned int num_parameters = _sampler.getNumberOfCols();
+
+  for (unsigned int i = 0; i < num_parameters; ++i)
```

Comment:
```suggestion
  for (const auto i : make_range(_sampler.getNumberOfCols())
```

Should be able to use this in lots of places

Reply:
Actually, I reverted the change since I use num_parameters within the loop.